pisa/stages/aeff/param.py

```diff
@@ -0,0 +1,140 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.'''
+
+
+def load_aeff_param(source):
+    """Load aeff parameterisation (energy- or coszen-dependent) from file
+    or dictionary.
+    Parameters
+    ----------
+    source : string or mapping
+        Source of the parameterization. If string, treat as file path or
+        resource location and load from the file; this must yield a mapping. If
+        `source` is a mapping, it is used directly. See notes below on format.
+    Returns
+    -------
+    aeff_params : OrderedDict
+        Keys are stringified flavintgroups and values are the callables that
+        produce aeff when called with energy or coszen values.
+    Notes
+    -----
+    The mapping passed via `source` or loaded therefrom must have the format:
+        {
+            <flavintgroup_string>: val,
+            <flavintgroup_string>: val,
+            ...
+        }
+    Note that the `transform_groups` (container.name) defined
+    in a pipeline config file using this must match the groupings defined
+    above.
+    `val`s can be one of the following:
+        - Callable with one argument
+        - String such that `eval(val)` yields a callable with one argument
+    """
+    if not (source is None or isinstance(source, (str, Mapping))):
+        raise TypeError('`source` must be string, mapping, or None')
+
+    if isinstance(source, str):
+        orig_dict = from_file(source)
+    elif isinstance(source, Mapping):
+        orig_dict = source
+    else:
+        raise TypeError('Cannot load aeff parameterizations from a %s'
+                        % type(source))
+
+    return orig_dict
+
+
+class param(Stage):
+    """Effective area service based on parameterisation functions stored in a
+    .json file.
+    Transforms an input map of a flux of a given flavour into maps of
+    event rates for the two types of weak current (charged or neutral),
+    according to energy and cosine zenith dependent effective areas specified
+    by parameterisation functions.
+    Parameters
+    ----------
+    params : ParamSet
+        Must exclusively have parameters:
+        aeff_energy_paramfile
+        aeff_coszen_paramfile
+        livetime
+        aeff_scale
+    """
+    def __init__(
+        self,
+        **std_kwargs,
+    ):
+        expected_params = (
+            'aeff_energy_paramfile',
+            'aeff_coszen_paramfile',
+            'livetime', 
+            'aeff_scale'
+        )
+
+        super(self.__class__, self).__init__(
+            expected_params=expected_params,
+            **std_kwargs,
+        )
+        
+        energy_param_source = self.params.aeff_energy_paramfile.value
+        coszen_param_source = self.params.aeff_coszen_paramfile.value
+        
+        self.energy_param = load_aeff_param(energy_param_source)
+        self.coszen_param = load_aeff_param(coszen_param_source)
+
+    @profile
```

Comment:
Is it encouraged to decorate `apply_function` like this, or rather to make use of the mechanism in https://github.com/icecube/pisa/blob/master/pisa/core/stage.py#L363?

Reply:
Good question. I have seen the @profile multiple times in the code, not sure what would be the better option.

Reply:
Here, if I'm not mistaken, whenever you run a pipeline with verbosity set to debug level, timing will be reported for the apply function... In the other case, since the mechanism applies to the base class itself and to setup, compute, and apply functions, we don't have to rely on decorators being present in each service to get timing information.

Reply:
Hence, I suggest removing the decorator if you have no strong reason for including it.